src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/ArraySinglePrimitiveRecord.cs

```diff
@@ -0,0 +1,212 @@
+    where T : unmanaged
+{
+    private static TypeName? s_elementTypeName;
+
+    internal ArraySinglePrimitiveRecord(ArrayInfo arrayInfo, IReadOnlyList<T> values) : base(arrayInfo)
+    {
+        Values = values;
+        ValuesToRead = 0; // there is nothing to read anymore
+    }
+
+    public override RecordType RecordType => RecordType.ArraySinglePrimitive;
+
+    public override TypeName ElementTypeName
+        => s_elementTypeName ??= TypeName.Parse(typeof(T).FullName.AsSpan()).WithAssemblyName(typeof(T).GetAssemblyNameIncludingTypeForwards());
+
+    internal IReadOnlyList<T> Values { get; }
+
+    public override bool IsTypeNameMatching(Type type) => typeof(T[]) == type;
+
+    internal override bool IsElementType(Type typeElement) => typeElement == typeof(T);
+
+    /// <inheritdoc/>
+    public override T[] GetArray(bool allowNulls = true)
+        => Values is T[] array ? array : Values.ToArray();
+
+    internal override (AllowedRecordTypes allowed, PrimitiveType primitiveType) GetAllowedRecordType()
+    {
+        Debug.Fail("GetAllowedRecordType should never be called on ArraySinglePrimitiveRecord");
+        throw new InvalidOperationException();
+    }
+
+    private protected override void AddValue(object value)
+    {
+        Debug.Fail("AddValue should never be called on ArraySinglePrimitiveRecord");
+        throw new InvalidOperationException();
+    }
+
+    internal static IReadOnlyList<T> DecodePrimitiveTypes(BinaryReader reader, int count)
+    {
+        if (typeof(T) == typeof(byte) && reader.IsDataAvailable(count))
+        {
+            T[] result = (T[])(object)reader.ReadBytes(count);
+            // This might be less than the number of bytes requested if the end of the stream is reached.
+            if (result.Length != count)
+            {
+                ThrowHelper.ThrowEndOfStreamException();
+            }
+            return result;
+        }
+        else if (typeof(T) == typeof(char) && reader.IsDataAvailable(count))
+        {
+            T[] result = (T[])(object)reader.ReadChars(count);
+            if (result.Length != count)
+            {
+                ThrowHelper.ThrowEndOfStreamException();
+            }
+            return result;
+        }
+
+        // Most of the tests use MemoryStream or FileStream and they both allow for executing the fast path.
+        // To ensure the slow path is tested as well, the fast path is executed only for optimized builds.
+#if NETCOREAPP
+        if (typeof(T) != typeof(decimal) && typeof(T) != typeof(DateTime) && typeof(T) != typeof(TimeSpan) // not optimized
+            && reader.IsDataAvailable(count * Unsafe.SizeOf<T>()))
+        {
+            return DecodePrimitiveTypesToArray(reader, count);
+        }
+#endif
+        return DecodePrimitiveTypesToList(reader, count);
+    }
+
+    private static List<T> DecodePrimitiveTypesToList(BinaryReader reader, int count)
+    {
+        List<T> values = [];
```

Comment:
This will grow from 0 to 4 to 8 to ...

The 99% case for large arrays will be bytes, and that is covered for most inputs by the "if I have that much data remaining, just read it" above.  But large arrays of TimeSpan (or int when not on new.NET) are going to have to go through many reallocs.

The easiest, of course, is

```suggestion
        List<T> values = new List<T>(count);
```

If the threat model indicates that this puts too much power in the hands of the payload, then it needs to change to a page based allocation and a custom type.

For expediency, I recommend just pre-sizing the list and taking this as an issue to follow up on later (as part of, or prior to, the threat model).  As current usage is limited to resources, it's a fine thing to do.

Reply:
The list is not pre-sized on purpose: https://github.com/adamsitnik/SafePayloadReader/issues/4

Since the code for decoding is already optimized for .NET Core, I am going to add the optimization for other TFMs